File: 2025/day-04/main2.py
def solve_printing_dept(filename):
    try:
        with open(filename, 'r') as f:
            lines = [line.strip() for line in f.readlines() if line.strip()]
    except FileNotFoundError:
        print(f"Error: Could not find file '{filename}'")
        return

    # Convert to a 2D grid (list of lists) so we can modify it
    grid = [list(line) for line in lines]
    rows = len(grid)
    cols = len(grid[0])

    # Directions for 8 neighbors: (row_change, col_change)
    neighbor_offsets = [
        (-1, -1), (-1, 0), (-1, 1),
        (0, -1),           (0, 1),
        (1, -1),  (1, 0),  (1, 1)
    ]

    print(f"Analyzing {rows}x{cols} grid for recursive removal...")

    total_removed = 0
    iteration = 0

    while True:
        iteration += 1
        rolls_to_remove = []

        # Step 1: Scan the grid to find all accessible rolls in this current state
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == '@':
                    neighbor_paper_count = 0

                    # Check all 8 neighbors
                    for dr, dc in neighbor_offsets:
                        nr, nc = r + dr, c + dc

                        # Boundary check
                        if 0 <= nr < rows and 0 <= nc < cols:
                            if grid[nr][nc] == '@':
                                neighbor_paper_count += 1

                    # The Rule: Forklifts can access if FEWER than 4 paper neighbors
                    if neighbor_paper_count < 4:
                        rolls_to_remove.append((r, c))

        # Step 2: If nothing to remove, we are done
        if not rolls_to_remove:
            break

        # Step 3: Remove the rolls (update grid state for next iteration)
        count_this_round = len(rolls_to_remove)
        total_removed += count_this_round

        # print(f"Iteration {iteration}: Removing {count_this_round} rolls.")

        for r, c in rolls_to_remove:
            grid[r][c] = '.' # Mark as empty space

    print("-" * 30)
    print(f"Total rolls of paper removed: {total_removed}")

if __name__ == "__main__":
    solve_printing_dept('input.txt')

File: 2025/day-04/main2.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func solvePrintingDeptPart2(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Printf("Error: Could not find file '%s'\n", filename)
		return
	}
	defer file.Close()

	// Read non-empty lines
	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		if line != "" {
			lines = append(lines, line)
		}
	}

	if len(lines) == 0 {
		fmt.Println("No data to process.")
		return
	}

	// Convert to mutable 2D grid
	rows := len(lines)
	cols := len(lines[0])
	grid := make([][]byte, rows)
	for i := 0; i < rows; i++ {
		grid[i] = []byte(lines[i])
	}

	// 8-direction neighbor offsets
	neighborOffsets := [][2]int{
		{-1, -1}, {-1, 0}, {-1, 1},
		{0, -1}, {0, 1},
		{1, -1}, {1, 0}, {1, 1},
	}

	fmt.Printf("Analyzing %dx%d grid for recursive removal...\n", rows, cols)

	totalRemoved := 0
	iteration := 0

	for {
		iteration++
		var rollsToRemove [][2]int

		// Step 1: Scan grid for accessible rolls
		for r := 0; r < rows; r++ {
			for c := 0; c < cols; c++ {
				if grid[r][c] == '@' {
					neighborPaperCount := 0

					for _, off := range neighborOffsets {
						nr := r + off[0]
						nc := c + off[1]

						if nr >= 0 && nr < rows && nc >= 0 && nc < cols {
							if grid[nr][nc] == '@' {
								neighborPaperCount++
							}
						}
					}

					// Rule: accessible if fewer than 4 neighbors
					if neighborPaperCount < 4 {
						rollsToRemove = append(rollsToRemove, [2]int{r, c})
					}
				}
			}
		}

		// Step 2: stop if nothing to remove
		if len(rollsToRemove) == 0 {
			break
		}

		// Step 3: remove rolls
		totalRemoved += len(rollsToRemove)

		// fmt.Printf("Iteration %d: Removing %d rolls.\n", iteration, len(rollsToRemove))

		for _, pos := range rollsToRemove {
			grid[pos[0]][pos[1]] = '.'
		}
	}

	fmt.Println("------------------------------")
	fmt.Printf("Total rolls of paper removed: %d\n", totalRemoved)
}

func main() {
	solvePrintingDeptPart2("input.txt")
}
